### PR TITLE
ELY-2731 Use SecureRandom instead of ThreadLocalRandom

### PR DESCRIPTION
**Pull request by TomasHofman**
Issue: https://issues.redhat.com/browse/ELY-2731

### CODE REVIEW

File: password/impl/src/test/java/org/wildfly/security/password/impl/PasswordUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.wildfly.security.password.impl;
```
Comment (darranl) [inline password/impl/src/test/java/org/wildfly/security/password/impl/PasswordUtilTest.java:18]: This is missing a copyright header.

File: password/impl/src/main/java/org/wildfly/security/password/impl/PasswordUtil.java [hunk 2]
```diff
@@ -35,7 +37,24 @@ final class PasswordUtil {
...
+     *
+     * @return a byte array representing the random salt
+     */
+    public static int generateRandomSaltInt() {
+        byte[] saltBytes = generateRandomSalt(4);
+        return convertBytesToInt(saltBytes);
+    }
+
+    static int convertBytesToInt(byte[] saltBytes) {
+        if (saltBytes.length != 4) {
```
Comment (darranl) [inline password/impl/src/main/java/org/wildfly/security/password/impl/PasswordUtil.java]: Can we do something here to use an Assert from wildfly-common?

Having an exception makes me feel this needs to be moved to a message / logger class but TBH as a method internal to this class we know it is unlikely to be anything other than 4 in length so an Assert call sounds safer to protect for the future.

File: password/impl/src/main/java/org/wildfly/security/password/impl/PasswordUtil.java [hunk 3]
```diff
@@ -35,7 +37,24 @@ final class PasswordUtil {
...
+        RANDOM.nextBytes(randomSalt);
         return randomSalt;
     }
+
+    /**
+     * Generate a random salt as int.
+     *
+     * @return a byte array representing the random salt
+     */
+    public static int generateRandomSaltInt() {
```
Comment (darranl) [inline password/impl/src/main/java/org/wildfly/security/password/impl/PasswordUtil.java]: I know the original was public but maybe reduce the visibility here as only used within this package.

**Comment by ivassile** [comment]
@TomasHofman Can you make the requested changes? Thanks!

**Comment by TomasHofman** [comment]
Sorry, lost track of this. I updated the PR.

One thing - I'm looking at the wildfly-common `Assert.assertTrue()`  method:

```
    public static boolean assertTrue(boolean expr) {
        assert expr : CommonMessages.msg.expectedBoolean(expr);
        return expr;
    }
```

and then the `CommonMessages.msg.expectedBoolean()` method:

```
    @Message(id = 1003, value = "Internal error: Assertion failure: Expected boolean value to be %s")
    String expectedBoolean(boolean expr);
```

It looks to me the printed message is wrong, and prints the opposite of the expected value. In the `assertTrue()`  method we should call `CommonMessages.msg.expectedBoolean(true);` rather than `CommonMessages.msg.expectedBoolean(expr);`..?

Or maybe it's just to early for me.

**Comment by darranl** [comment]
Looking at the JavaDoc for SecureRandom, yes the class itself is thread safe and can be accessed by multiple threads concurrently:

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/SecureRandom.html

BUT if the underlying Provider does not declare it is thread safe it will introduce locking.on accessing the provider.

I don't know if it will be better or worse but I wonder if Elytron could have it's on ThreadLocalSecureRandom - the reason for the ThreadLocal model is you know your instance is only going to be accessed by a single thread so you are not waiting on locks held by other threads.  The down side being you end up with one SecureRandom per thread but is this an issue?

**Comment by TomasHofman** [comment]
I updated the PR - added a publicly accessible ThreadLocalSecureRandom class.

I guess since the ThreadLocal model was there in the first place, it's probably safer to stick to it.

I think you are able to asses the performance implication better than me. My *feeling* originally was that the random bytes generation in practice would not be as heavily used for this to cause performance issues, but whether that's true or not, there's no reason to introduce bottlenecks that do not need to be there.

File: password/impl/src/main/java/org/wildfly/security/password/impl/ThreadLocalSecureRandom.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.password.impl;
+
+import java.security.SecureRandom;
+import java.util.function.Supplier;
+
+public class ThreadLocalSecureRandom implements Supplier<SecureRandom> {
```
Comment (darranl) [inline password/impl/src/main/java/org/wildfly/security/password/impl/ThreadLocalSecureRandom.java]: This looks great thank you, one further minor change - could you please drop the public modifier as this is only accessed from classes in the same package.

Later we may want to move to a more central location but for now this is great for these classes.

**Comment by darranl** [comment]
@ivassile if you can add a review I think with one more change from @TomasHofman I can merge today

File: password/impl/src/main/java/org/wildfly/security/password/impl/ThreadLocalSecureRandom.java (hunk 4 above)
Comment (TomasHofman) [reply password/impl/src/main/java/org/wildfly/security/password/impl/ThreadLocalSecureRandom.java]: Sure, I made it package local.

**Review by darranl — APPROVED** [review]

**Comment by darranl** [comment]
I have replaced this with https://github.com/wildfly-security/wildfly-elytron/pull/2189 to get us some better CI results.

**Pull request closed without merging**